### PR TITLE
Create a python controller that can be used with pio

### PR DESCRIPTION
**Pull request by cmcantalupo**

### CODE REVIEW

File: service/geopmdpy/session.py [hunk 1]
```diff
@@ -44,7 +44,7 @@
 from dasbus.error import DBusError
 from . import topo
 from . import pio
-
+from . import runtime
```
Comment (bgeltz) [inline service/geopmdpy/session.py:47]:
```suggestion
from . import runtime

```

File: service/geopmdpy/runtime.py [hunk 2]
```diff
@@ -115,3 +118,114 @@ def __next__(self):
...
+        pio.save_control()
+        try:
+            for ss in self._signals:
+                pio.push_signal(*ss)
+            for cc in self._controls:
+                pio.push_control(*cc)
+            pio.read_batch()
+            signals_begin = self.read_all_signals()
+            pid = subprocess.Popen(self._argv)
+            for loop_idx in TimedLoop(self._num_update, self._update_period):
```
Comment (bgeltz) [inline service/geopmdpy/runtime.py]: Use 0 to mean infinite loop?

File: service/geopmdpy/runtime.py [hunk 3]
```diff
@@ -115,3 +118,114 @@ def __next__(self):
...
+                    break
+                pio.read_batch()
+                signals = self.read_all_signals()
+                controls = self._agent.update(signals)
+                for control_idx in range(len(self._controls)):
+                    pio.adjust(control_idx, control[control_idx])
+                pio.write_batch()
+            pio.read_batch()
+            signals_end = self.read_all_signals()
+            result = [ee - bb for bb, ee in zip(signals_begin, signals_end)]
```
Comment (bgeltz) [inline service/geopmdpy/runtime.py]: A reporter method would be nice.

Comment (bgeltz) [reply service/geopmdpy/runtime.py]: result = agent.generateReport()

File: service/test/test_controller.py [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+#  (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY LOG OF THE USE
+#  OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+#
+
+import geopmdpy.runtime
+import geopmdpy.topo
+
+class LocalAgent(geopmdpy.runtime.Agent):
+    def get_signals(self):
+        return [("INSTRUCTIONS_RETIRED", geopmdpy.topo.DOMAIN_PACKAGE, 0)]
```
Comment (bgeltz) [inline service/test/test_controller.py]: Including TIME would help.

File: service/test/test_controller.sh [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+#  LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+#  DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+#  THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+#  (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY LOG OF THE USE
+#  OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+#
+
+python3 test_controller.py
+
+exit 0
```
Comment (bgeltz) [inline service/test/test_controller.sh]: Delete this file.

File: service/Makefile.am [hunk 6]
```diff
@@ -72,6 +72,8 @@ EXTRA_DIST = .gitignore \
              test/README.md \
              test/build_dasbus.sh \
              test/install_service.sh \
+             test/test_controller.py \
+             test/test_controller.sh \
```
Comment (bgeltz) [inline service/Makefile.am]:
```suggestion
```

**Review by bgeltz — CHANGES_REQUESTED** [review]

**Review by bgeltz — APPROVED** [review]

**Comment by alawibaba** [comment]
I had to make a couple of small changes to get through the test. Also since we took away the sh script (which admittedly I forgot to do correctly), should we dump a usage message if parameters are absent?

File: service/test/test_controller.py [hunk 7]
```diff
@@ -66,10 +72,24 @@ def get_report(self):
...
+    err = 0
+    help_msg = f"""\
+Usage: {sys.argv[0]} COMMAND
+
+{LocalAgent.__doc__}
+    """
+    if len(sys.argv) < 2:
+        err = -1
+    if err != 0 or sys.argv[1] == '--help':
+        sys.stderr.print(help_msg)
```
Comment (cmcantalupo) [inline service/test/test_controller.py]: print -> write...testing is tbd

**Review by bgeltz — APPROVED** [review]

**Pull request merged**